skimage/future/graph/graph_cut.py

```diff
@@ -122,20 +121,88 @@ def cut_normalized(labels, rag, thresh=0.001, num_cuts=10, in_place=True,
+    """Generate normalized graph cuts on the region adjacency graph.
+
+    Given an image's labels and its similarity RAG, recursively perform
+    a two-way normalized cut on it. All nodes belonging to a subgraph
+    that cannot be cut further are assigned a unique label in the
+    output. Appending a new value to the thresh list and calling
+    next on this generator returns a new array of labels updated
+    for the new thresh value.
+
+    Parameters
+    ----------
+    labels : ndarray
+        The array of labels.
+    rag : RAG
+        The region adjacency graph.
+    init_thresh : float
+        Initial threshold value to segment with. A subgraph won't be further
+        subdivided if the value of the N-cut exceeds a threshold value.
+    num_cuts : int
+        The number or N-cuts to perform before determining the optimal one.
+    in_place : bool
+        If set, modifies `rag` in place. For each node `n` the function will
+        set a new attribute ``rag.node[n]['ncut label']``.
+    max_edge : float, optional
+        The maximum possible value of an edge in the RAG. This corresponds to
+        an edge between identical regions. This is used to put self
+        edges in the RAG.
+
+    Yields
+    ------
+    out : ndarray
+        The new labeled array.
+
+    Examples
+    --------
+    >>> from skimage import data, segmentation
+    >>> from skimage.future import graph
+    >>> img = data.astronaut()
+    >>> labels = segmentation.slic(img, compactness=30, n_segments=400)
+    >>> rag = graph.rag_mean_color(img, labels, mode='similarity')
+    >>>
+    >>> init_thresh = 1e-5
+    >>> new_label_gen = graph.cut_normalized_gen(labels, rag, init_thresh)
+    >>> new_labels = next(new_label_gen)
+    >>> new_labels = new_label_gen.send(1e-3)
+    >>> new_labels = new_label_gen.send(1e-4)
+
+    References
+    ----------
+    .. [1] Shi, J.; Malik, J., "Normalized cuts and image segmentation",
+           Pattern Analysis and Machine Intelligence,
+           IEEE Transactions on, vol. 22, no. 8, pp. 888-905, August 2000.
+
+    """
     if not in_place:
         rag = rag.copy()
 
     for node in rag.nodes():
         rag.add_edge(node, node, weight=max_edge)
 
-    _ncut_relabel(rag, thresh, num_cuts)
+    # Initialize the generator
+    ncut_gen = _ncut_relabel(rag, 0, num_cuts)
+    next(ncut_gen)
 
-    map_array = np.zeros(labels.max() + 1, dtype=labels.dtype)
-    # Mapping from old labels to new
-    for n, d in rag.nodes(data=True):
-        map_array[d['labels']] = d['ncut label']
+    # Generate new labels as thresholds come
+    thresh = init_thresh
+    while True:
+        ncut_gen.send(thresh)
 
-    return map_array[labels]
+        map_array = np.zeros(labels.max() + 1, dtype=labels.dtype)
+        # Mapping from old labels to new
+        for n, d in rag.nodes(data=True):
+            map_array[d['labels']] = d['ncut label']
+
+        thresh = yield map_array[labels]
```

Comment:
If I'm not mistaken, calling `next` on an initialized generator will result in a nasty error, as `thresh` will be set to `None`. Perhaps we can add some kind of check to ensure that `thresh` is valid? I think most users won't be familiar with the `send` interface (it's actually the first time I see it used outside of a contrived tutorial!), so we need to be wary of misuse, more so than in normal functions.

Reply:
Send is pretty unfamiliar to me too. Raising an error on None is probably a good idea.